### ISSUE TITLE
toString/equality discrepancy for Uris with trailing slashes

### ISSUE THREAD

**Issue by ceedubs** [http4s/http4s#676]
If two `Uri`s are equivalent except for a trailing slash then they are not considered equal, but it's tough to diagnose how they differ because their `toString` and `renderString` methods return the same value. See the example below.

I'm not entirely sure in which direction this should be changed (to make their string version match or change their equality implementation). I do think it's confusing behavior that was worth flagging though.

```scala
scala> import org.http4s.Uri.uri
import org.http4s.Uri.uri

scala> val withSlash = uri("http://foo.bar/")
withSlash: org.http4s.Uri = http://foo.bar

scala> val withoutSlash = uri("http://foo.bar")
withoutSlash: org.http4s.Uri = http://foo.bar

scala> withSlash == withoutSlash
res0: Boolean = false

scala> withSlash.renderString == withoutSlash.renderString
res1: Boolean = true

scala> withSlash.toString == withoutSlash.toString
res2: Boolean = true
```

**Comment by rossabaker** [comment]
Yeah, this bit recently, too.  Thanks for bringing it up.

Does anybody have a strong opinion which way this should be unified?  @refried and @bryce-anderson have spent the most time in the URI RFC muck.

**Comment by refried** [comment]
I think it should render with the slash if it was created with the slash.

**Comment by refried** [comment]
This probably means making the path portion optional if it isn’t already?  I can’t remember whether that will have other annoying implications though.  Maybe it won’t :)

**Comment by rossabaker** [comment]
I am sure that there is a semantic difference between `http://foo.bar/baz` and `http://foo.bar/baz/`.  The fact that this is special cased makes me think we decided there's no semantic difference between `http://foo.bar` and `http://foo.bar/`.  And this might be one of those answers that varies by the http protocol vs. generic URI schemes.

**Comment by refried** [comment]
Yeah, or may even vary as common web app implementation vs HTTP spec!  Sorry for short answers, am on mobile.

**Comment by rossabaker** [comment]
This seems to be the [special case](https://github.com/http4s/http4s/blob/162962ff91d184703d5daaffbafc1573089ba02e/core/src/main/scala/org/http4s/Uri.scala#L80).  I don't see any justification for it in the RFC.  Does anybody remember why this is here?

**Comment by rossabaker** [comment]
Probably relates to this, from section 6.2.3 of RFC3986:

> For example, because the "http" scheme makes use of an authority component, has a default port of "80", and defines an empty path to be equivalent to "/", the following four URIs are equivalent:

>> http://example.com
>> http://example.com/
>> http://example.com:/
>> http://example.com:80/

> In general, a URI that uses the generic syntax for authority with an empty path should be normalized to a path of "/".  Likewise, an explicit ":port", for which the port is empty or the default for the scheme, is equivalent to one where the port and its ":" delimiter are elided and thus should be removed by scheme-based normalization.  For example, the second URI above is the normal form for the "http" scheme.

This raises the question of how far to go into normalization when determining equality.  This is an HTTP-specific rule, but one to be followed "in general", and we're primarily an HTTP library.  There are myriad other complicated normalization rules that we could consider in equality, and don't.  I'm inclined to say URI equality should be lexical instead of canonical, and it _mostly_ is at this point.  But if we go that far, is the case insensitivity of schemes still a good thing to do?

**Comment by rossabaker** [comment]
This is what java.net.URI does:

>If the given object is not a URI then this method immediately returns false.

> For two URIs to be considered equal requires that either both are opaque or both are hierarchical. Their schemes must either both be undefined or else be equal without regard to case. Their fragments must either both be undefined or else be equal.

> For two opaque URIs to be considered equal, their scheme-specific parts must be equal.

> For two hierarchical URIs to be considered equal, their paths must be equal and their queries must either both be undefined or else be equal. Their authorities must either both be undefined, or both be registry-based, or both be server-based. If their authorities are defined and are registry-based, then they must be equal. If their authorities are defined and are server-based, then their hosts must be equal without regard to case, their port numbers must be equal, and their user-information components must be equal.

> When testing the user-information, path, query, fragment, authority, or scheme-specific parts of two URIs for equality, the raw forms rather than the encoded forms of these components are compared and the hexadecimal digits of escaped octets are compared without regard to case.

This goes through 6.2.2.1 of the equivalence "ladder" specified by RFC 3986.

**Comment by rossabaker** [comment]
The "Identities" section of the [Uri javadoc](http://download.java.net/jdk7/archive/b123/docs/api/java/net/URI.html#equals(java.lang.Object)) also suggests some laws that could be useful if we ever had an `Arbitrary[Uri]`.  Including the one that made @ceedubs itch.

**Comment by refried** [comment]
If we wanted to, down the line, we could do scheme-specific normalization during construction?  What do you think?

**Comment by rossabaker** [comment]
Scheme-specific normalization is a huge rabbit hole, but in an HTTP library, normalizing 'http://' URIs could certainly be in scope.  I'm also not sure that it should be on construction as opposed to a function on the `Uri`.  Whether such normalizations should affect equality is another discussion to be had, with that RFC ladder as a reference.

**Comment by rossabaker** [comment]
Fixed by #683
